topic-operator/src/main/java/io/strimzi/operator/topic/LocalStoreConfiguration.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.topic;
+
+import io.apicurio.registry.utils.streams.diservice.AsyncBiFunctionService;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StoreQueryParameters;
+import org.apache.kafka.streams.state.QueryableStoreTypes;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+
+/**
+ * Simple local / in-memory store configuration which is sufficient when the TO runs in a single pod
+ */
+class LocalStoreConfiguration implements StoreConfiguration {
+    private ReadOnlyKeyValueStore<String, Topic> store;
+    private BiFunction<String, String, CompletionStage<Integer>> lookupService;
+
+    @Override
+    public ReadOnlyKeyValueStore<String, Topic> getStore() {
+        return store;
+    }
+
+    @Override
+    public BiFunction<String, String, CompletionStage<Integer>> getLookupService() {
+        return lookupService;
+    }
+
+    @SuppressWarnings("unchecked")
+    public void configure(
+            Config config,
+            Properties kafkaProperties,
+            KafkaStreams streams,
+            AsyncBiFunctionService.WithSerdes<String, String, Integer> serviceImpl,
+            List<AutoCloseable> closeables
+    ) {
+        String storeName = config.get(Config.STORE_NAME);
+        // we need to lazily create the store as streams might not be ready yet
+        store = (ReadOnlyKeyValueStore<String, Topic>) Proxy.newProxyInstance(
+                getClass().getClassLoader(),
+                new Class[]{ReadOnlyKeyValueStore.class},
+                new InvocationHandler() {
+                    private ReadOnlyKeyValueStore<String, Topic> store;
+
+                    @Override
+                    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+                        if (store == null) {
+                            synchronized (this) {
+                                if (store == null) {
+                                    store = streams.store(StoreQueryParameters.fromNameAndType(storeName, QueryableStoreTypes.keyValueStore()));
```

Comment:
AFAICS this is plain old double checked locking. I don't see how it can be thread safe (if I'm wrong can you explain). But I rather suspect that it's not worth trying to optimize here and just making it `synchronized invoke()` would give adequate performance and be correct from a thread safety pov.

Reply:
This is classic double lock to handle concurrency, and you don't need full synchronized on the method.
But yeah, no need to optimize this, will just use plain sync on the method.

Reply:
AFAIU (and I don't claim to be an expert) double checked locking doesn't work because although the reference to store is handled correctly it's still possible that the state of the allocated object get published unsafely. I don't what `streams.store()` does, but assuming it's not synchronized I think you have to assume that it returns a new object and so the construction overall is not safe. This is why I think just making the whole thing `synchronized` makes sense here, because it safes us from having to reason about whether DCL is really safe here.